mapogcsld.c

```diff
@@ -1455,50 +1466,81 @@ int msSLDParseOgcExpression(CPLXMLNode *psRoot, void *psObj, int binding,
           case MS_LABEL_BASE + MS_LABEL_BINDING_OUTLINECOLOR:
             strDelim = "\"";
           default:
-            propertyString = msStringConcatenate(propertyString, strDelim);
-            propertyString = msStringConcatenate(propertyString, "[");
-            propertyString = msStringConcatenate(propertyString, psRoot->psChild->pszValue);
-            propertyString = msStringConcatenate(propertyString, "]");
-            propertyString = msStringConcatenate(propertyString, strDelim);
+            msStringBufferAppend(property, strDelim);
+            msStringBufferAppend(property, "[");
+            msStringBufferAppend(property, psRoot->psChild->pszValue);
+            msStringBufferAppend(property, "]");
+            msStringBufferAppend(property, strDelim);
             msInitExpression(&(psStyle->exprBindings[binding]));
-            exprBindings[binding].string = propertyString;
+            exprBindings[binding].string =
+                msStringBufferReleaseStringAndFree(property);
             exprBindings[binding].type = MS_EXPRESSION;
             (*nexprbindings)++;
             break;
         }
         status = MS_SUCCESS;
       }
+      else if (strcasecmp(psRoot->pszValue,"Function") == 0
+          && CPLGetXMLValue(psRoot,"name",NULL)
+          && psRoot->psChild->psNext)
+      {
+        // Parse a <ogc:Function> element
+        msStringBuffer * function = msStringBufferAlloc();
+
+        // Parse function name
+        const char * funcname = CPLGetXMLValue(psRoot,"name",NULL);
+        msStringBufferAppend(function, funcname);
+        msStringBufferAppend(function, "(");
+        msInitExpression(&(exprBindings[binding]));
+
+        // Parse arguments
+        char * sep ="";
+        for (CPLXMLNode * argument = psRoot->psChild->psNext ; argument ; argument = argument->psNext)
+        {
+          status = msSLDParseOgcExpression(argument, psObj, binding, objtype);
+          if (status != MS_SUCCESS)
+            break;
+          msStringBufferAppend(function, sep);
+          msStringBufferAppend(function, exprBindings[binding].string);
+          msFree(exprBindings[binding].string);
+          msInitExpression(&(exprBindings[binding]));
+          sep = ",";
+        }
+        msStringBufferAppend(function, ")");
+        exprBindings[binding].string =
+            msStringBufferReleaseStringAndFree(function);
+        exprBindings[binding].type = MS_EXPRESSION;
+        (*nexprbindings)++;
+        status = MS_SUCCESS;
+      }
       else if (strstr(ops, psRoot->pszValue)
           && psRoot->psChild && psRoot->psChild->psNext)
       {
         // Parse an arithmetic element <ogc:Add>, <ogc:Sub>, <ogc:Mul>, <ogc:Div>
         const char operator[2] = { *(strstr(ops, psRoot->pszValue)+3), '\0' };
-        char * expressionString = NULL;
+        msStringBuffer * expression = msStringBufferAlloc();
```

Comment:
may potentially leak if msStringBufferReleaseStringAndFree() is not reached

Reply:
Good catch, thank you!